client/src/Services/socketClient.ts

```diff
@@ -0,0 +1,217 @@
+  type: 'error';
+  message: string;
+  code?: number;
+}
+
+export interface SocketPackageMainDebug extends SocketPackageBase {
+  type: 'mainDebug';
+  debugInfo: string;
+}
+
+export interface SocketPackageExecuteNext extends SocketPackageBase {
+  type: 'executeNext';
+  command: string;
+}
+
+export interface SocketPackageSendDummyLinkedListData extends SocketPackageBase {
+  type: 'sendDummyLinkedListData';
+  linkedList: any[];
+}
+
+export interface SocketPackageSendDummyBinaryTreeData extends SocketPackageBase {
+  type: 'sendDummyBinaryTreeData';
+  binaryTree: any;
+}
+
+export interface SocketPackageSendFunctionDeclaration extends SocketPackageBase {
+  type: 'sendFunctionDeclaration';
+  functionCode: string;
+}
+
+export interface SocketPackageSendTypeDeclaration extends SocketPackageBase {
+  type: 'sendTypeDeclaration';
+  typeDef: string;
+}
+
+export interface SocketPackageSendBackendStateToUser extends SocketPackageBase {
+  type: 'sendBackendStateToUser';
+  state: any;
+}
+
+export interface SocketPackageSendStdoutToUser extends SocketPackageBase {
+  type: 'sendStdoutToUser';
+  output: string;
+}
+
+export interface SocketPackageProgramWaitingForInput extends SocketPackageBase {
+  type: 'programWaitingForInput';
+  inputRequired: boolean;
+}
+
+export interface SocketPackageCompileError extends SocketPackageBase {
+  type: 'compileError';
+  errors: string[];
+}
+
+export interface SocketPackageSendStdin extends SocketPackageBase {
+  type: 'send_stdin';
+  inputData: string;
+}
+
+export interface SocketPackageConnectError extends SocketPackageBase {
+  type: 'connect_error';
+}
+
+export type SocketPackageConcrete =
+  | SocketPackageConnect
+  | SocketPackageDisconnect
+  | SocketPackageSendData
+  | SocketPackageReceiveData
+  | SocketPackageError
+  | SocketPackageMainDebug
+  | SocketPackageExecuteNext
+  | SocketPackageSendDummyLinkedListData
+  | SocketPackageSendDummyBinaryTreeData
+  | SocketPackageSendFunctionDeclaration
+  | SocketPackageSendTypeDeclaration
+  | SocketPackageSendBackendStateToUser
+  | SocketPackageSendStdoutToUser
+  | SocketPackageProgramWaitingForInput
+  | SocketPackageCompileError
+  | SocketPackageSendStdin
+  | SocketPackageConnectError;
+
+type ServerToClientEvents = {
+  [s in SocketEventType]: SocketPackageConcrete;
+};
+
+interface ClientToServerEvents {
+  mainDebug: (code: string) => void;
+}
+
+class SocketClient {
+  private socket: Socket<ServerToClientEvents, ClientToServerEvents>;
+
+  get socketTemp(): Socket {
+    return this.socket;
+  }
+
+  constructor() {
+    this.socket = io(URL);
+  }
+
+  attach() {}
+
+  connect() {
+    this.socket.on('connect', () => {
+      console.log('Connected!');
+    });
+  }
+
+  disconnect() {
+    this.socket.on('disconnect', () => {
+      console.log('Disconnected!');
+    });
+  }
+
+  on<T extends SocketEventType>(event: T, callback: (data: SocketPackageConcrete) => void) {
```

Comment:
Suggestion:

```typescript
 on<EventType extends keyof ServerToClientEvents>(
    event: EventType,
    callback: (data: ServerToClientEvents[EventType]) => void
  ) {
```
Better because links each specific `SocketEventType` to corresponding `SocketPackageConcrete`
e.g `on(event: "sendData", callback: (data: SocketPackageSendData) => void) {...}`